### PR TITLE
Fix wrong diagnostic parse typed pattern

### PR DESCRIPTION
**Pull request by kimdv**
While digging in #1484 I found this.

### CODE REVIEW

File: Tests/SwiftParserTest/translated/RecoveryTests.swift [hunk 1]
```diff
@@ -869,6 +876,23 @@ final class RecoveryTests: XCTestCase {
...
+      private var a 1️⃣b : Int = ""
+      """,
+      diagnostics: [
+        DiagnosticSpec(
+            locationMarker: "1️⃣",
+            message: "unexpected code 'b' before type annotation"
+        )
+      ],
+      fixedSource: """
+        private var a b : Int = ""
```
Comment (kimdv) [inline Tests/SwiftParserTest/translated/RecoveryTests.swift:891]: Should we in this case merge `a` and `b`

Comment (ahoppen) [reply Tests/SwiftParserTest/translated/RecoveryTests.swift:891]: I think that would make sense.

Comment (kimdv) [reply Tests/SwiftParserTest/translated/RecoveryTests.swift]: Fixed and rebased in main

**Comment by kimdv** [comment]
@swift-ci please test

File: Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift [hunk 2]
```diff
@@ -1414,6 +1414,36 @@ public class ParseDiagnosticsGenerator: SyntaxAnyVisitor {
...
+      return .skipChildren
+    }
+
+    if let identifierPattern = node.pattern.as(IdentifierPatternSyntax.self),
+      let unexpectedBeforeColon = node.typeAnnotation?.unexpectedBeforeColon,
+      let token = unexpectedBeforeColon.onlyToken(where: { $0.tokenKind.isIdentifier })
+    {
+      addDiagnostic(
+        unexpectedBeforeColon,
+        SpaceSeparatedIdentifiersError(
```
Comment (ahoppen) [inline Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift:1428]: Do you know why we need to do special handling here and why this is not caught by the general logic in `visit(UnexpectedNodesSyntax)`?

Comment (kimdv) [reply Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift:1428]: The reason was that `previousToken.parent?.is(DeclSyntax.self) == true`. Parent here is an `IdentifierPatternSyntax`. 
I added the pattern as I think it's okay to merge an identifier with the previous identifier pattern.

**Comment by kimdv** [comment]
@swift-ci please test

**Comment by kimdv** [comment]
@swift-ci please test windows

**Review by ahoppen — APPROVED** [review]
Thanks. That’s much cleaner now

**Pull request merged**